crates/iceberg/src/spec/table_metadata.rs

```diff
@@ -278,6 +330,229 @@ impl TableMetadata {
         self.snapshots
             .insert(snapshot.snapshot_id(), Arc::new(snapshot));
     }
+
+    /// Normalize this partition spec.
+    ///
+    /// This is an internal method
+    /// meant to be called after constructing table metadata from untrusted sources.
+    /// We run this method after json deserialization.
+    /// All constructors for `TableMetadata` which are part of `iceberg-rust`
+    /// should return normalized `TableMetadata`.
+    ///
+    /// It does:
+    /// * Validate the current schema is set and valid
+    /// * Validate that all refs are valid (snapshot exists)
+    /// * Validate logs are chronological
+    /// * Normalize location (remove trailing slash)
+    /// * Validate that for V1 Metadata the last_sequence_number is 0
+    /// * If the default partition spec is specified but the spec is not present in specs, add it
+    /// * If the default sort order is unsorted but the sort order is not present, add it
+    pub(super) fn try_normalize(&mut self) -> Result<&mut Self> {
+        self.validate_current_schema()?;
+        self.normalize_current_snapshot()?;
+        self.validate_refs()?;
+        self.validate_chronological_snapshot_logs()?;
+        self.validate_chronological_metadata_logs()?;
+        self.location = self.location.trim_end_matches('/').to_string();
+        self.validate_format_version_specifics()?;
+        self.try_normalize_partition_spec()?;
+        self.try_normalize_sort_order()?;
+        Ok(self)
+    }
+
+    fn try_normalize_partition_spec(&mut self) -> Result<()> {
```

Comment:
Hi, I'm not sure if this makes the logic easier to read. (It works for me, though).

```rust
if self.partition_spec_by_id(self.default_spec_id).is_some() {
   return Ok(())
}

if self.default_spec_id != DEFAULT_PARTITION_SPEC_ID {
  return Err(...)
}

let partition_spec = xxx;
self.partition_specs.insert(xxx);
Ok(())
```

The `self.partition_specs.is_empty()` check can be avoided since it should never occur.